sql-plugin/src/main/scala/com/nvidia/spark/rapids/NvcompLZ4CompressionCodec.scala

```diff
@@ -0,0 +1,169 @@
+    val tempSize = LZ4Compressor.getTempSize(input, CompressionType.CHAR, LZ4_CHUNK_SIZE)
+    withResource(DeviceMemoryBuffer.allocate(tempSize)) { tempBuffer =>
+      var compressedSize: Long = 0L
+      val outputSize = LZ4Compressor.getOutputSize(input, CompressionType.CHAR, LZ4_CHUNK_SIZE,
+        tempBuffer)
+      closeOnExcept(DeviceMemoryBuffer.allocate(outputSize)) { outputBuffer =>
+        compressedSize = LZ4Compressor.compress(input, CompressionType.CHAR, LZ4_CHUNK_SIZE,
+          tempBuffer, outputBuffer, stream)
+        require(compressedSize <= outputBuffer.getLength, "compressed buffer overrun")
+        (compressedSize, outputBuffer)
+      }
+    }
+  }
+
+  /**
+   * Decompress data asynchronously that was compressed with nvcomp's LZ4-GPU codec
+   * @param outputBuffer where the uncompressed data will be written
+   * @param inputBuffer buffer of compressed data to decompress
+   * @param stream CUDA stream to use
+   */
+  def decompressAsync(
+      outputBuffer: DeviceMemoryBuffer,
+      inputBuffer: DeviceMemoryBuffer,
+      stream: Cuda.Stream): Unit = {
+    withResource(Decompressor.getMetadata(inputBuffer, stream)) { metadata =>
+      val outputSize = Decompressor.getOutputSize(metadata)
+      if (outputSize != outputBuffer.getLength) {
+        throw new IllegalStateException(
+          s"metadata uncompressed size is $outputSize, buffer size is ${outputBuffer.getLength}")
+      }
+      val tempSize = Decompressor.getTempSize(metadata)
+      withResource(DeviceMemoryBuffer.allocate(tempSize)) { tempBuffer =>
+        Decompressor.decompressAsync(inputBuffer, tempBuffer, metadata, outputBuffer, stream)
+      }
+    }
+  }
+}
+
+class BatchedNvcompLZ4Compressor(maxBatchMemorySize: Long, stream: Cuda.Stream)
+    extends BatchedTableCompressor(maxBatchMemorySize, stream) {
+  override protected def compress(
+      tables: Array[ContiguousTable],
+      stream: Cuda.Stream): Array[CompressedTable] = {
+    val inputBuffers: Array[BaseDeviceMemoryBuffer] = tables.map(_.getBuffer)
+    val compressionResult = BatchedLZ4Compressor.compress(inputBuffers,
+      NvcompLZ4CompressionCodec.LZ4_CHUNK_SIZE, stream)
+    val compressedTables = try {
+      val buffers = compressionResult.getCompressedBuffers
+      val compressedSizes = compressionResult.getCompressedSizes
+      buffers.zipWithIndex.map { case (buffer, i) =>
+        val contigTable = tables(i)
+        val compressedSize = compressedSizes(i)
+        require(compressedSize <= buffer.getLength, "compressed buffer overrun")
+        val meta = MetaUtils.buildTableMeta(
+          tableId = 0,
+          contigTable.getTable,
+          contigTable.getBuffer,
+          CodecType.NVCOMP_LZ4,
+          compressedSize)
+        CompressedTable(compressedSize, meta, buffer)
+      }
+    } catch {
+      case t: Throwable =>
+        compressionResult.getCompressedBuffers.safeClose()
```

Comment:
would be nice if `CompressionResult` was `AutoCloseable`, that way we don't have to reach in it to clean its state in this case.

Reply:
since this is in cudf, certainly can be followup and is minor.

Reply:
I explicitly did not make it `AutoCloseable` because the results are expected to be adopted by the caller.  If `CompressionResult` was `AutoCloseable` then that would imply one would want to use it in a try-with-resources which would be actively harmful in the typical use-case.

Reply:
ok sounds reasonable. I do wonder if have a convenience method to clean any type of native bits that `CompressionResult` could contain in case of failure would be good to have, but I don't think is an issue right now.